### ISSUE TITLE
factory.create(): allow any name/values for initialization

### ISSUE THREAD

**Issue by lharder** [defold/defold#7058]
**Problem**: When I create gameobjects, I usually want to specify their attributes upon initialization, e.g. to have a common behavior but different properties (speed, health, firepower, you name it). Today, this is possible by providing a table with properties to factory.create(). 

Properties, however, for good reasons, allow only certain variable types. Strings, for example, are not permitted. Neither are more complex data structures like tables. This is because properties of game objects are exposed and can be referenced by other game objects, so performance is an issue. 

However, there is no reason why one should have to expose all initialization values publicly for the entire lifetime of a game object. Here, Defold mixes up two aspects that have no logical connection.

There are many scenarios, where I need to initialize the game object with data without wanting to expose it to other game objects. An obvious example would be a name to be displayed in dialogs. Or, for a complex type of spaceship, it would be useful to provide data on many parts of the vessel in a single table for its engines, weapons, etc. There is no need, however, for other game objects to access these - actually, by default, it would be wise to prevent that.

**Suggestion**: factory.create() already provides the possibility to pass a table with names/values upon creation of a game object which must be defined as properties. I suggest to:

(a) remove the restriction that only the values of properties can be passed. Any name/value pairs should be allowed. It should be possible to have both properties and non-properties mixed in the same table. For non-properties, complex data types (like string and table) should be permitted. 

(b) Extend the signature of the game objects' init method and include the table with names/values from factory.create() as a method parameter:

```
-- outside the game object ---------
local id = factory.create( "#myfactory", nil, nil, { speed = 100,  caption = "Bob"  } )

-- inside the game object --------------
go.property( "speed", 0 )

function init( self, attrs )
  self.displayname = attrs[ caption ]
  
  pprint( self.displayname )	-- "Bob"
  pprint( self.speed )		-- 100
end
```

That way, in the init method, the game object has full access to anything it gets passed in, no matter if property or not. Init can receive strings, tables, as well as simpler properties. 

Outside of the init method, nothing changes: properties getting passed in upon factory.create() remain as they are today and can be referenced from the outside. Names/values not declared as properties are local to init() only, they are not accessible from the outside and so there is no danger of performance issues. 

**Alternatives**: There are workarounds for this problem, of course:

**(a) Global variable**: It is possible to define a table, make it globally available to all game objects and pass in the right key as a property to perform a lookup in the game object's init() method:

```
-- outside the game object ---------
GLOBALS = {}
GLOBALS[ 1 ] = { caption = "Bob" }
local id = factory.create("#myfactory", nil, nil, { key = 1  } )

-- game object --------------
go.property( "key", 0 )
go.property( "speed", 0 )

function init( self, attrs )
  self.displayname = GLOBALS[ self.key ].caption
  
  pprint( self.displayname )	-- "Bob"
  pprint( self.speed )		-- 100
end
```

However, there are several downsides to this approach:

1. Global variables should be avoided if possible. To need them usually is an indicator of bad code.
2. Details of the game object must be exposed (read and write) via the global data structure although they may only be needed by a single game object. 
3. The game object must declare and expose an otherwise useless property "key" just to look up the values required upon initialization
4. The developer must do more housekeeping: after factory.create(), the data in the global variable is no longer needed, but still it is available to all(!) other parts of the code

**(b) Message passing**: Right after the game object has been created, one can send a message with an arbitrary table as payload and the game object can wait for and evaluate that message. However, this is an even more cumbersome workaround:

1. Upon initialization, the data is not available yet. Instead of in the init() method, where logically, the game object should be setup, initialization code is dispersed.

2. Instead, the update method must perform a check on every single frame for the rest of the game object's lifetime and must perform code which, by design, should not be in this method at all:

```
function init( self )
  -- required data is not available
  -- can only do other initialization code
end

function on_message( self, message_id, message )
  -- wait for the right message to come
  -- then parse and setup data as required
end

function update( self, dt )
  -- worst part: check on every(!) frame if init data has been sent and is available
  -- only act if object has been initialized properly, do not act before
end
```

Both workarounds are cumbersome and ugly.

The change I suggest would provide elegance and flexibility for a large number of usecases while remaining downward compatible with existing code.

As for the effort to implement this behavior, I assume it must be limited as no new concepts get introduced and the only requirement would be to remove a seemingly arbitrary restriction concerning data types and to extend the init() method signature with an optional parameter that is available already.

**Comment by britzl** [comment]
When you call `factory.create()` there is a lot of stuff happening. Especially if you also provide a properties table. These values are serialized and passed to the created game object components as messages. We do not want to open up for arbitrary sized data on game object creation and we want to limit the data types to a set of primitive types (number, bool) and some Defold specific ones (has, vector3 etc).

Our strong recommendation is to look up associated and complex data belonging to a game object using some kind of key, typically the game object id (hash). To use your example:

```
-- lookup.lua
-- this module tracks object state/data associated to ids (hash)
-- it's possible to put, get and remove data
local M = {}

local states = {}

function M.put(id, data)
  assert(id)
  assert(data)
  states[id] = data
end

function M.get(id)
  id = id or go.get_id()
  return states[id]
end

function M.remove(id)
  id = id or go.get_id()
  states[id] = nil
end

return M
```

```
-- spawner.script
-- this is a script that spawns game objects
-- the script also associates some data with the spawned object
local states = require "lookup"

function init(self)
  local data = {
    caption = "Bob",
  }

  local id = factory.create("#myfactory")
  states.put(id, data)
end
```

```
-- object.script
-- this is the spawned object
-- it looks up associated data using its own id
local states = require "lookup"

function init(self)
  local data = states.get(go.get_id())
  local data = states.get() -- alternative
  print(data.caption)
end

function final(self)
  states.remove(go.get_id())
  states.remove() -- alternative
end
```

**Comment by lharder** [comment]
Hi Britzl,

thanks for your comment and suggestion. That would be a workaround I could live with - unfortunately, it does not work as you describe it, if I am not mistaken?

Using the newly created game object's id as its key to a globally available data structure would spare the introduction of that clumsy "key" property. That would be fair enough for me. Unfortunately, there is a timing problem (unless I get something completely wrong, but I just tried it in code and that confirms the problem):

You suggest: 
````
-- outside the game object
local id = factory.create("#myfactory")  -- executed 1st
states.put(id, data)			 -- executed 4th

-- inside the game object
function init(self)
  local data = states.get( go.get_id() ) -- executed 2nd
  print( data.caption )			 -- executed 3rd
end
````

The problem is that, when you call factory.create(), you do not yet know the new game object's id - it is returned only after the call. So, when you put the data into the table with the newly generated id as its key, the init() method has already finished. It finds no data, because the structure at that point is still empty. 

Inside the init() method in your code example, data.caption is nil. Is it not?

Your suggestion would work if the newly spawned game object were initialized only the next frame, so that states.put(id, data) would get executed before init(). But when I try it, that does not seem to be the case? 

I would not mind keeping properties and complex data separated in two different parameters if that makes more sense to you. At any rate, that would avoid having to pass complex data on to all components. 

But unless I am mistaken, without that suggested feature we are stuck with only clumsy workarounds for a very common problem. Or do I overlook something?

**Comment by britzl** [comment]
Oh, yes, I was typing that up on the fly and didn't think about the init order. My apologies! Delaying the code in init can be 
done in two ways, neither is particularly beautiful, but also not super ugly:

```
function init(self)
	-- use a timer
	timer.delay(0, false, function()
		local data = states.get(go.get_id())
		print(data.caption)
	end)

	-- use a message
	msg.post("#", "delayed_init")
end


function on_message(self, message_id, message, sender)
	if message_id == hash("delayed_init") then
		local data = states.get(go.get_id())
		print(data.caption)
	end
end
```

**Comment by AGulev** [comment]
I would recommend to implement it like this:

my_configs.lua
```
local M ={}

local configs = {}

--some configs predefined
configs[hash("tank")] = {["Caption"] = "Tank", speed = 2}

--add new configs
function M.put(type, data)
  assert(type)
  assert(data)
  configs[type] = data
end

function M.get(type)
  return configs[type]
end

function M.remove(type)
  configs[type] = nil
end

return M
```

Object script:
```
local configs = require "my_configs"

go.property( "type", hash("empty") )

function init( self )
  self.config  = configs.get(self.type)
...
end

```

Script where you spawn the objects:
```
local configs = require "my_configs"

function init()
    local type = hash("my_new_type")
    configs.put(type, {caption = "Bob", speed = 20})
    local id = factory.create("#myfactory", nil, nil,  { type = type  }) 
end
```

You did the same with globals, yes, but it would be better to have a config file where you can pre-define different object types when you work on balance and so on, but it also has an ability to add new configs in runtime.

I think it's a proper way to do it.

**Comment by lharder** [comment]
Thanks everybody for the discussion and the different suggestions. 

Personally, I am still convinced that the additional initialisation parameter for local-only, complex data would be valuable and more elegant than any of the workarounds. But I concede that this may be a matter of personal taste.

I appreciate your feedback.

**Comment by britzl** [comment]
Note: we're actually looking into a new kind of data component that could potentially solve this kind of issue.

**Comment by lharder** [comment]
Sounds awesome - you have just made me very, very curious :) 
Looking forward to anything of the kind!

**Comment by britzl** [comment]
> you have just made me very, very curious :)

It will be a continuation of the [simple data component](https://github.com/defold/extension-simpledata), but more flexible and integrated into the engine.
